File: lib/ui/charts/chart_empty_state.dart
import 'package:flutter/material.dart';

/// What a chart says when there is nothing to draw.
///
/// The app had four answers. `MetricDayChart` wrote a muted sentence. The caffeine
/// curve wrote a muted sentence of its own, in its own file. `MetricSessionChart`
/// and `MetricLineChart` returned `SizedBox.shrink()` — they do not say "no data",
/// they VANISH, and the card around them vanishes with them, so the screen silently
/// reflows and you are left to notice that a thing you were looking for is not
/// there.
///
/// This is the one, and in this commit it renders exactly what the sentence-writers
/// already rendered — same style, same colour, same words — so nothing moves. It
/// exists to be a single place to change, because "give the charts a real empty
/// state" is a job that must be done once and is otherwise done four times and
/// badly.
///
/// The vanishing charts are NOT converted here. Turning "absent" into "an empty
/// card" changes the layout of the heart, vitals and activity screens — a card
/// appears where there was none — and that is a user-visible behaviour change, not
/// a refactor. It gets its own commit, with its own screen tests.
class ChartEmptyState extends StatelessWidget {
  const ChartEmptyState({
    super.key,
    required this.message,
    this.height,
  });

  final String message;

  /// Reserve the chart's footprint, so a card that resolves from empty to full
  /// does not make the page jump. Null keeps the current behaviour: the text takes
  /// only the room it needs.
  final double? height;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final scheme = theme.colorScheme;

    final content = Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        // Muted, and small. An empty state is the least important thing on the
        // screen: it should be legible and then get out of the way, not sit there
        // as a large grey exclamation about a day you simply did not log.
        Icon(
          Icons.show_chart,
          size: 28,
          color: scheme.onSurfaceVariant.withValues(alpha: 0.45),
        ),
        const SizedBox(height: 8),
        Text(
          message,
          textAlign: TextAlign.center,
          style: theme.textTheme.bodyMedium
              ?.copyWith(color: scheme.onSurfaceVariant),
        ),
      ],
    );

    if (height == null) {
      return Padding(
        padding: const EdgeInsets.symmetric(vertical: 12),
        child: Center(child: content),
      );
    }
    return SizedBox(
      height: height,
      width: double.infinity,
      child: Center(child: content),
    );
  }
}

File: app/src/main/kotlin/tech/mmarca/openvitals/ui/charts/ChartEmptyState.kt
package tech.mmarca.openvitals.ui.components

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ShowChart
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/** What a chart says with nothing to draw. [height] reserves the footprint so the page does not jump. */
@Composable
fun ChartEmptyState(
    message: String,
    modifier: Modifier = Modifier,
    height: Dp? = null,
) {
    val sized = if (height == null) {
        modifier
            .fillMaxWidth()
            .padding(vertical = 12.dp)
    } else {
        modifier
            .fillMaxWidth()
            .height(height)
    }
    Box(modifier = sized, contentAlignment = Alignment.Center) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            // Muted and small: an empty state should get out of the way.
            Icon(
                imageVector = Icons.AutoMirrored.Outlined.ShowChart,
                contentDescription = null,
                modifier = Modifier.size(28.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.45f),
            )
            Spacer(Modifier.height(8.dp))
            Text(
                text = message,
                textAlign = TextAlign.Center,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}
